### PR TITLE
Get GMP building on Apple Silicon

### PR DESCRIPTION
**Pull request by BytesGuy**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

This is a bit of a halfway fix to get GMP building and working on Apple Silicon. The architecture is not yet supported officially, so while we wait this will produce a generic C build - according to the documentation this may run slowly, but at least it gets it working for now!

### CODE REVIEW

**Comment by claui** [comment]
This might be really helpful, given that lots of popular formulae depend on gmp.

**Comment by BytesGuy** [comment]
> This might be really helpful, given that lots of popular formulae depend on gmp.

Yeah, it's probably about as good as it gets for now I started working on this because I was trying to install GCC which depends on it - working on that next!

**Comment by fxcoudert** [comment]
Alternatively there is a patch for this on the gmp-bugs mailing list: https://gmplib.org/list-archives/gmp-bugs/2020-July/004837.html

**Comment by BytesGuy** [comment]
> Alternatively there is a patch for this on the gmp-bugs mailing list: https://gmplib.org/list-archives/gmp-bugs/2020-July/004837.html
> It would enable the use of aarch64 assembly, making GMP more efficient

I have just been trying to get this to work with little success for some reason. Are you doing anything noteworthy that I might be missing here? Fairly new to this so apologies if it's something obvious!

Applying the patch and having no `--build` defined for configure provides:

```
==> Patching
==> Applying attachment.bin
patching file configure.ac
patching file mpn/arm64/arm64-defs.m4
patching file mpn/arm64/bdiv_q_1.asm
patching file mpn/arm64/darwin.m4
patching file mpn/arm64/invert_limb.asm
==> ./configure
Last 15 lines from /Users/adam/Library/Logs/Homebrew/gmp/01.configure:
checking if globals are prefixed by underscore... yes
checking how to switch to read-only data section... 	.section	__TEXT,__const
checking for assembler .type directive... 
checking for assembler .size directive... 
checking for assembler local label prefix... L
checking for assembler byte directive... .byte
checking how to define a 32-bit word... .long
checking if .align assembly directive is logarithmic... yes
checking size of void *... 8
checking size of unsigned short... 2
checking size of unsigned... 4
checking size of unsigned long... 8
checking size of mp_limb_t... 8
configure: error: Oops, mp_limb_t is 64 bits, but the assembler code
in this configuration expects 32 bits.
/usr/bin/curl --version
/usr/libexec/java_home --xml --failfast
/usr/local/Homebrew/Library/Homebrew/debrew.rb:13:in `raise'
```

Adding just `--build=aarch64-apple-darwin20` to configure provides:

```
==> Patching
==> Applying attachment.bin
patching file configure.ac
patching file mpn/arm64/arm64-defs.m4
patching file mpn/arm64/bdiv_q_1.asm
patching file mpn/arm64/darwin.m4
patching file mpn/arm64/invert_limb.asm
==> ./configure --build=aarch64-apple-darwin20
==> make
Last 15 lines from /Users/adam/Library/Logs/Homebrew/gmp/02.make:
       ^
tmp-bdiv_q_1.s:76:12: error: expected ')' in parentheses expression
 LEA_LO( x7, ___gmp_binvert_limb_table)
           ^
tmp-bdiv_q_1.s:76:8: error: invalid operand
 LEA_LO( x7, ___gmp_binvert_limb_table)
       ^
make[2]: *** [bdiv_q_1.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
libtool: compile:  clang -DHAVE_CONFIG_H -I. -I.. -D__GMP_WITHIN_GMP -I.. -DOPERATION_dcpi1_bdiv_qr -O2 -pedantic -march=armv8-a -c dcpi1_bdiv_qr.c  -fno-common -DPIC -o .libs/dcpi1_bdiv_qr.o
libtool: compile:  clang -DHAVE_CONFIG_H -I. -I.. -D__GMP_WITHIN_GMP -I.. -DOPERATION_sbpi1_bdiv_r -O2 -pedantic -march=armv8-a -c sbpi1_bdiv_r.c -o sbpi1_bdiv_r.o >/dev/null 2>&1
libtool: compile:  clang -DHAVE_CONFIG_H -I. -I.. -D__GMP_WITHIN_GMP -I.. -DOPERATION_dcpi1_bdiv_q -O2 -pedantic -march=armv8-a -c dcpi1_bdiv_q.c -o dcpi1_bdiv_q.o >/dev/null 2>&1
libtool: compile:  clang -DHAVE_CONFIG_H -I. -I.. -D__GMP_WITHIN_GMP -I.. -DOPERATION_dcpi1_bdiv_qr -O2 -pedantic -march=armv8-a -c dcpi1_bdiv_qr.c -o dcpi1_bdiv_qr.o >/dev/null 2>&1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
/usr/bin/curl --version
/usr/libexec/java_home --xml --failfast
/usr/local/Homebrew/Library/Homebrew/debrew.rb:13:in `raise'
```

**Comment by fxcoudert** [comment]
Since it's a configure.ac patch, it needs: 

`system "autoreconf", "-fiv"`

and build-dependencies on `autoconf`, `automake`, and `libtool`.

**Comment by claui** [comment]
@BytesGuy Whenever a patch modifies `configure.ac`, we need to run `autoreconf`. [Example](https://github.com/Homebrew/homebrew-core/blob/628584be5a80ddf9b5d46c87063070299f616dcb/Formula/libopendkim.rb#L30-L31)

TIL!

**Comment by claui** [comment]
With `autoreconf` in place, the formula builds now but the static linking test fails. https://github.com/Homebrew/homebrew-core/blob/dc859fcb2dbf2cfb4085322903be411e5b9b9b27/Formula/gmp.rb#L50-L52

```plain
Undefined symbols for architecture arm64:
  "__gmpn_pi1_bdiv_q_1", referenced from:
      ___gmpn_bdiv_q_1 in libgmp.a(bdiv_q_1.o)
     (maybe you meant: ___gmpn_pi1_bdiv_q_1)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

It’s hard to see with some fonts but somehow it seems to be about the number of leading `_`s, again.

**Comment by BytesGuy** [comment]
@fxcoudert @claui Thank you for the pointers! 🙂 I really need to brush up on these things again, it's been a long time! Just pushed a new commit which now seems to build ok. Tests fail though, as mentioned. Hopefully this is at least a step in the right direction 🤞

**Comment by claui** [comment]
I think the branch statement in `bdiv_q_1.asm` is what’s causing the error:

```
LEA_HI(	x7, binvert_limb_table)
ubfx	x6, d, 1, 7
LEA_LO(	x7, binvert_limb_table)
ldrb	w6, [x7, x6]
ubfiz	x7, x6, 1, 8
umull	x6, w6, w6
msub	x6, x6, d, x7
lsl	x7, x6, 1
mul	x6, x6, x6
msub	x6, x6, d, x7
lsl	x7, x6, 1
mul	x6, x6, x6
msub	di, x6, d, x7

b	mpn_pi1_bdiv_q_1    ←
```

<details>
  <summary>Evidence</summary>
  <pre>$ nm /usr/local/lib/libgmp.a | grep gmpn_pi 
                 U ___gmpn_pi1_bdiv_q_1
                 U ___gmpn_pi1_bdiv_q_1
                 U ___gmpn_pi1_bdiv_q_1
                 U ___gmpn_pi1_bdiv_q_1
                 U ___gmpn_pi1_bdiv_q_1
                 U ___gmpn_pi1_bdiv_q_1
0000000000000048 T ___gmpn_pi1_bdiv_q_1
                 U __gmpn_pi1_bdiv_q_1    ←
                 U ___gmpn_pi1_bdiv_q_1
                 U ___gmpn_pi1_bdiv_q_1

$ nm /usr/local/lib/libgmp.a | grep -A 6 bdiv_q_1.o
/usr/local/lib/libgmp.a(bdiv_q_1.o):
                 U ___gmp_binvert_limb_table
0000000000000000 T ___gmpn_bdiv_q_1
0000000000000048 T ___gmpn_pi1_bdiv_q_1
                 U __gmpn_pi1_bdiv_q_1    ←
0000000000000000 t ltmp0
</pre>
</details>

**Review by BytesGuy — COMMENTED** [review]
Awesome, thank you! 🎉

**Comment by claui** [comment]
The CI failures in `adios2`, `agda` and `ncview` are obviously unrelated.

**Review by claui — DISMISSED** [review]

**Comment by claui** [comment]
Pinging @fxcoudert for a second review/opinion.

File: Formula/gmp.rb
Comment (fxcoudert) [inline Formula/gmp.rb:29]:
```suggestion
  depends_on "autoconf"  => :build if Hardware::CPU.arm?
```

and same below

File: Formula/gmp.rb
Comment (fxcoudert) [inline Formula/gmp.rb:22]: Consolidate the patches in a formula-patches file

File: Formula/gmp.rb
Comment (fxcoudert) [inline Formula/gmp.rb:44]: Why are `--enable-cxx --with-pic` not passed on ARM? These should not be platform-dependent.

**Review by fxcoudert — CHANGES_REQUESTED** [review]
See comments

File: Formula/gmp.rb
Comment (BytesGuy) [reply Formula/gmp.rb:44]: @fxcoudert passing these caused failures for me when producing a generic C build and before adding the patch. Can re-test this soon to confirm whether or not this is still the case.

Comment (BytesGuy) [reply Formula/gmp.rb]: Confirmed this works since including the patch

File: Formula/gmp.rb
Comment (BytesGuy) [reply Formula/gmp.rb]: Tidied up depends in 7b1d75b

**Comment by claui** [comment]
Double-checked that the consolidated patch works on arm64.

Waiting on CI (style/audit only) to pass, and for https://github.com/Homebrew/formula-patches/pull/288 to be merged.

**Comment by glyng1** [comment]
Good afternoon,

I am currently having problems with my andriod and computer.

So sorry regards Glyn 😢

File: Formula/gmp.rb
Comment (claui) [inline Formula/gmp.rb:21]: Needs update due to https://github.com/Homebrew/formula-patches/pull/288 being rebased and merged.

```suggestion
    url "https://raw.githubusercontent.com/Homebrew/formula-patches/c53834b4/gmp/6.2.0-arm.patch"
```

**Review by claui — APPROVED** [review]

**Comment by claui** [comment]
Thanks to all of you! This is a textbook example for excellent collaboration among contributors and with upstream. ❤️

**Comment by BytesGuy** [comment]
> Thanks to all of you! This is a textbook example for excellent collaboration among contributors and with upstream. ❤️

Thank you for working on this too! So glad to see this fixed as it should unblock a number of other packages 🎉

**Pull request merged**
